### PR TITLE
ref: Refactor MetricRule to be TS

### PR DESCRIPTION
**Pull request by ryan953**

### CODE REVIEW

File: static/app/views/alerts/types.tsx [hunk 1]
```diff
@@ -87,7 +87,18 @@ export enum CombinedAlertType {
   ISSUE = 'rule',
 }
 
-export type CombinedMetricIssueAlerts = (IssueAlertRule | MetricRule) & {
-  type: CombinedAlertType;
+// export type CombinedMetricIssueAlerts = (IssueAlertRule | MetricRule) & {
+//   type: CombinedAlertType;
+//   latestIncident?: Incident | null;
+// };
```
Comment (evanpurkhiser) [inline static/app/views/alerts/types.tsx:93]: comment?

File: static/app/views/alerts/types.tsx [hunk 2]
```diff
@@ -87,7 +87,18 @@ export enum CombinedAlertType {
...
+
+interface IssueAlert extends IssueAlertRule {
+  type: CombinedAlertType.ISSUE;
   latestIncident?: Incident | null;
```
Comment (scttcper) [inline static/app/views/alerts/types.tsx:97]: issue alerts don't have incidents

File: fixtures/js-stubs/metricRule.ts [hunk 3]
```diff
@@ -1,10 +1,16 @@
+import {Dataset} from 'sentry/views/alerts/rules/metric/types';
+import {CombinedAlertType, CombinedMetricIssueAlerts} from 'sentry/views/alerts/types';
+
 import {IncidentTrigger} from './incidentTrigger';
 
-export function MetricRule(params = {}) {
+export function MetricRule<
+  T extends CombinedMetricIssueAlerts & {type: CombinedAlertType.METRIC}
+>(params: Partial<T> = {}): CombinedMetricIssueAlerts {
```
Comment (scttcper) [inline fixtures/js-stubs/metricRule.ts:8]: I'm not sure this makes sense to do just to get the alert list tests to pass

File: static/app/views/alerts/types.tsx [hunk 4]
```diff
@@ -87,7 +87,18 @@ export enum CombinedAlertType {
...
-};
+}
+interface MetricAlert extends MetricRule {
+  type: CombinedAlertType.METRIC;
+  latestIncident?: Incident | null;
+}
+
+export type CombinedMetricIssueAlerts = IssueAlert | MetricAlert;
```
Comment (scttcper) [inline static/app/views/alerts/types.tsx:104]: this type is better 👍

File: static/app/views/alerts/types.tsx (hunk 1 above)
Comment (ryan953) [reply static/app/views/alerts/types.tsx:93]: oops

File: fixtures/js-stubs/metricRule.ts (hunk 3 above)
Comment (ryan953) [reply fixtures/js-stubs/metricRule.ts:8]: it's a `MetricRule` generator, and the type `CombinedMetricIssueAlerts` allows for two types of rules, so imo it's proper to scope things down, the return value doesn't need to be either-or

File: static/app/views/alerts/types.tsx (hunk 2 above)
Comment (ryan953) [reply static/app/views/alerts/types.tsx:97]: I ported it over from the old thing:

```
export type CombinedMetricIssueAlerts = (IssueAlertRule | MetricRule) & {
  type: CombinedAlertType;
  latestIncident?: Incident | null;
};
```

Comment (ryan953) [reply static/app/views/alerts/types.tsx]: updated that type 🤞

File: fixtures/js-stubs/metricRule.ts (hunk 3 above)
Comment (scttcper) [reply fixtures/js-stubs/metricRule.ts:8]: i mean having the `type: CombinedAlertType.` since this was made to return a metric alert object

**Review by scttcper — APPROVED** [review]

**Pull request merged**